File: app/local_plugins/device_jieli/android/src/main/kotlin/com/jielihome/jielihome/feature/translation/TranslationModeHandler.kt
package com.jielihome.jielihome.feature.translation

import com.jieli.bluetooth.impl.JL_BluetoothManager

/**
 * 翻译模式处理器统一接口。每种模式独立实现：
 *  - 声明自己消费哪些「输入流」（采集面）和产出哪些「输出流」（回放面）
 *  - 实现 start：启动 SDK 端音频流（进入翻译模式 / 起手机麦），把解码后的 PCM 通过 bridge 推出
 *  - 实现 stop：关闭音频流，释放编解码器
 *  - 实现 onTranslatedAudio：接到外部翻译服务回送的 PCM，按 outputStreamId 路由
 */
interface TranslationModeHandler {
    val modeId: Int
    val inputStreams: List<String>
    val outputStreams: List<String>

    fun start(args: Map<String, Any?>)
    fun stop()
    val isWorking: Boolean

    /** 外部翻译服务回送的 PCM 注入点，由 TranslationFeature 路由到当前活动 handler */
    fun onTranslatedAudio(
        outputStreamId: String,
        pcm: ByteArray,
        format: AudioFormat,
        isFinal: Boolean,
    ): Boolean
}

abstract class BaseTranslationModeHandler(
    protected val btManager: JL_BluetoothManager,
    protected val bridge: TranslationAudioBridge,
) : TranslationModeHandler {

    @Volatile
    protected var working: Boolean = false

    override val isWorking: Boolean get() = working

    /** 把采集到的一帧 PCM 推给外部翻译服务 */
    protected fun pushFrame(
        streamId: String,
        pcm: ByteArray,
        format: AudioFormat = AudioFormat(),
        seq: Long = System.nanoTime(),
        isFinal: Boolean = false,
    ) {
        bridge.emitAudioFrame(modeId, streamId, pcm, format, seq, System.currentTimeMillis(), isFinal)
    }

    protected fun emitLog(content: String) = bridge.emitLog(modeId, content)
    protected fun emitError(code: Int, msg: String?) = bridge.emitError(modeId, code, msg)
}

/**
 * 与 SDK [com.jieli.bluetooth.bean.translation.TranslationMode] 完全一致的取值，
 * 方便直接拿插件 modeId 去构造 SDK 的 TranslationMode。
 */
object TranslationModeIds {
    const val MODE_IDLE = 0
    const val MODE_RECORD = 1
    const val MODE_RECORDING_TRANSLATION = 2
    const val MODE_CALL_TRANSLATION = 3
    const val MODE_AUDIO_TRANSLATION = 4
    const val MODE_FACE_TO_FACE_TRANSLATION = 5
    const val MODE_CALL_TRANSLATION_WITH_STEREO = 6
}

File: app/local_plugins/devices/device_jieli/ios/device_jieli/Sources/device_jieli/feature/translation/TranslationModeHandler.swift
import Flutter
import Foundation
import JL_BLEKit

/// 翻译模式处理器协议。与 Android `TranslationModeHandler` 对齐：
///   - 声明自己消费哪些「输入流」、产出哪些「输出流」
///   - `start` 进入翻译模式，启动上下行通路
///   - `stop` 退出模式，释放编解码器
///   - `onTranslatedAudio` 接外部 TTS PCM，按输出流路由
public protocol TranslationModeHandler: TranslationSessionOwner {
    var modeId: Int { get }
    var inputStreams: [String] { get }
    var outputStreams: [String] { get }
    var isWorking: Bool { get }
    func start(args: [String: Any], completion: @escaping (_ ok: Bool, _ errMsg: String?) -> Void)
    func stop()
    func onTranslatedAudio(streamId: String, pcm: Data, format: AudioFormat, isFinal: Bool) -> Bool
}

/// 公共基类：维护 `working` 标志，提供 pushFrame / emitLog / emitError 等工具。
public class BaseTranslationModeHandler: NSObject, TranslationModeHandler {

    public let modeId: Int
    public let inputStreams: [String]
    public let outputStreams: [String]

    weak var server: JieliHomeServer?
    var session: TranslationSession?
    @objc dynamic var working: Bool = false

    public var isWorking: Bool { working }

    public var sessionModeId: Int { modeId }

    private var seq: Int64 = 0

    public init(server: JieliHomeServer, modeId: Int, inputs: [String], outputs: [String]) {
        self.server = server
        self.modeId = modeId
        self.inputStreams = inputs
        self.outputStreams = outputs
        super.init()
    }

    // 子类需要重写
    public func start(args: [String: Any], completion: @escaping (Bool, String?) -> Void) {
        completion(false, "unimplemented")
    }
    public func stop() { working = false }
    public func onTranslatedAudio(streamId: String, pcm: Data, format: AudioFormat, isFinal: Bool) -> Bool { false }

    // TranslationSessionOwner
    public func onReceiveAudio(_ audio: JLTranslateAudio) {}
    public func onModeChange(_ mode: JLTranslateSetMode) {
        // 设备退出回到 IDLE 时收尾
        if mode.modeType == .idle && working {
            emitError(-1, "device exited mode → IDLE")
            stop()
        }
    }
    public func onError(_ error: NSError) { emitError(error.code, error.localizedDescription) }

    // MARK: - Helpers

    func pushFrame(
        streamId: String,
        pcm: Data,
        format: AudioFormat = AudioFormat(),
        isFinal: Bool = false,
        encoding: String = "pcm16"
    ) {
        seq &+= 1
        server?.dispatcher.send([
            "type": "translationAudio",
            "modeId": modeId,
            "streamId": streamId,
            "sampleRate": format.sampleRate,
            "channels": format.channels,
            "bitsPerSample": format.bitsPerSample,
            "encoding": encoding,
            "seq": NSNumber(value: seq),
            "tsMs": NSNumber(value: Int64(Date().timeIntervalSince1970 * 1000)),
            "final": isFinal,
            "pcm": FlutterStandardTypedData(bytes: pcm),
        ])
    }

    func emitLog(_ content: String) {
        server?.dispatcher.send([
            "type": "translationLog",
            "modeId": modeId,
            "content": content,
        ])
    }

    func emitError(_ code: Int, _ message: String?) {
        server?.dispatcher.send([
            "type": "translationError",
            "modeId": modeId,
            "code": code,
            "message": message as Any,
        ])
    }

    // MARK: - Mode helpers

    /// 解析常用枚举参数（audioType / strategy）
    static func parseAudioType(_ raw: Any?) -> JL_SpeakDataType? {
        if let s = raw as? String {
            switch s.lowercased() {
            case "opus": return .OPUS
            case "pcm": return .PCM
            case "speex": return .SPEEX
            case "jla_v2", "jlav2": return .JLA_V2
            case "msbc", "m_sbc": return .MSBC
            default: break
            }
        }
        if let n = (raw as? NSNumber)?.uint8Value {
            return JL_SpeakDataType(rawValue: n)
        }
        return nil
    }

    /// 解析录音策略（与 Android STRATEGY_DEVICE_ALWAYS_RECORDING / CUSTOM 对齐）
    static func parseRecordType(_ raw: Any?, default def: JLTranslateRecordType) -> JLTranslateRecordType {
        if let s = raw as? String {
            switch s.lowercased() {
            case "phone", "custom": return .byPhone
            case "device", "always": return .byDevice
            default: break
            }
        }
        if let n = (raw as? NSNumber)?.intValue {
            return JLTranslateRecordType(rawValue: n) ?? def
        }
        return def
    }

    /// 构造 SDK 的 JLTranslateSetMode
    static func buildMode(
        modeType: JLTranslateSetModeType,
        audioType: JL_SpeakDataType,
        channel: Int,
        sampleRate: Int
    ) -> JLTranslateSetMode {
        let m = JLTranslateSetMode()
        m.modeType = modeType
        m.dataType = audioType
        m.channel = channel
        m.sampleRate = sampleRate
        return m
    }
}
